backends/bmv2/action.cpp

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "action.h"
+
+namespace BMV2 {
+
+#ifdef PSA
+void
+ConvertActions::genExternMethod(Util::JsonArray* result, P4::ExternMethod *em) {
+    auto name = "_" + em->actualExternType->name + "_" + em->method->name.name;
+    auto primitive = mkPrimitive(name, result);
+    auto parameters = mkParameters(primitive);
+
+    auto ext = new Util::JsonObject();
+    ext->emplace("type", "extern");
+
+    // FIXME: PSA have extern pass building a map and lookup here.
+    if (em->object->is<IR::Parameter>()) {
+        auto param = em->object->to<IR::Parameter>();
+        // auto packageObject = resolveParameter(param);
+        // ext->emplace("value", packageObject->getName());
+        ext->emplace("value", "FIXME");
+    } else {
+        ext->emplace("value", em->object->getName());
+    }
+    parameters->append(ext);
+
+    for (auto a : *mc->arguments) {
+        auto arg = conv->convert(a);
+        parameters->append(arg);
+    }
+}
+#endif
+
+
+void
+ConvertActions::convertActionBody(const IR::Vector<IR::StatOrDecl>* body, Util::JsonArray* result) {
+    for (auto s : *body) {
+        // TODO(jafingerhut) - add line/col at all individual cases below,
+        // or perhaps it can be done as a common case above or below
+        // for all of them?
+        if (!s->is<IR::Statement>()) {
+            continue;
+        } else if (auto block = s->to<IR::BlockStatement>()) {
+            convertActionBody(&block->components, result);
+            continue;
+        } else if (s->is<IR::ReturnStatement>()) {
+            break;
+        } else if (s->is<IR::ExitStatement>()) {
+            auto primitive = mkPrimitive("exit", result);
+            (void)mkParameters(primitive);
+            primitive->emplace_non_null("source_info", s->sourceInfoJsonObj());
+            break;
+        } else if (s->is<IR::AssignmentStatement>()) {
+            const IR::Expression* l, *r;
+            auto assign = s->to<IR::AssignmentStatement>();
+            l = assign->left;
+            r = assign->right;
+
+            cstring operation;
+            auto type = typeMap->getType(l, true);
+            if (type->is<IR::Type_StructLike>())
+                operation = "copy_header";
+            else
+                operation = "modify_field";
```

Comment:
there is no mapping (also some new core primitives have no old equivalent), you'll have to go by the names
for `exit`, you should keep using `exit`, the name is staying the same